### PR TITLE
One connection cache hit

### PR DESCRIPTION
**Pull request by ashish-sharma09**

### CODE REVIEW

File: pipe-storage-postgresql/src/test/groovy/com/tesco/aqueduct/pipe/storage/ClusterStorageIntegrationSpec.groovy [hunk 1]
```diff
@@ -1,339 +0,0 @@
-package com.tesco.aqueduct.pipe.storage
```
Comment (LorenzoCianciaruso) [inline pipe-storage-postgresql/src/test/groovy/com/tesco/aqueduct/pipe/storage/ClusterStorageIntegrationSpec.groovy:1]: Moved to the integration folder

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 2]
```diff
@@ -63,35 +63,74 @@ public MessageResults read(
...
-        try (Connection connection = pipeDataSource.getConnection()) {
             LOG.info("getConnection:time", Long.toString(System.currentTimeMillis() - start));
 
-            connection.setAutoCommit(false);
-            setWorkMem(connection);
+            final Optional<ClusterCache> clusterCache = clusterStorage.getClusterCache(locationUuid, connection);
 
-            final long globalLatestOffset = offsetFetcher.getGlobalLatestOffset(connection);
+            if (clusterCache.isPresent() && clusterCache.get().isValidAndUnExpired()) {
+                return readMessages(types, start, startOffset, clusterCache.get().getClusterIds(), connection);
```
Comment (adam-yorwerth) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:75]: I think the naming of `clusterCache` here is a bit confusing, as we're only getting a cache entry for the location, not the whole cache. 

Perhaps we could go for `ClusterCacheEntry` ?

Comment (LorenzoCianciaruso) [reply pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:75]: yes, we weren't convinced by the naming either. I think the suggestion makes sense

**Review by sebastianlavender — APPROVED** [review]
LGTM

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 3]
```diff
@@ -63,35 +63,78 @@ public MessageResults read(
...
-            try (PreparedStatement messagesQuery = getMessagesStatement(connection, types, startOffset, globalLatestOffset, clusterIds)) {
+            } else {
+                close(connection);
 
-                final List<Message> messages = runMessagesQuery(messagesQuery);
-                long end = System.currentTimeMillis();
+                final List<String> clusterUuids = clusterStorage.resolveClustersFor(locationUuid);
 
-                final long retry = calculateRetryAfter(end - start, messages.size());
+                connection = pipeDataSource.getConnection();
```
Comment (adam-yorwerth) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:82]: Can we add a timing log around this too? Using the same 'getConnection:time'

Comment (adam-yorwerth) [reply pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:82]: Might be good to extract to another method even, with timing locally in the method

Comment (LorenzoCianciaruso) [reply pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:82]: yes, good point. I'm also gonna add a log when we recursively read

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 4]
```diff
@@ -63,35 +63,84 @@ public MessageResults read(
...
 
-                final long retry = calculateRetryAfter(end - start, messages.size());
+                final Optional<List<Long>> newClusterIds = clusterStorage.updateAndGetClusterIds(locationUuid, clusterUuids, entry, connection);
 
-                LOG.info("PostgresSqlStorage:retry", String.valueOf(retry));
-                return new MessageResults(messages, retry, OptionalLong.of(globalLatestOffset), PipeState.UP_TO_DATE);
+                if (newClusterIds.isPresent()) {
+                    return readMessages(types, start, startOffset, newClusterIds.get(), connection);
+                } else {
+                    LOG.info("postgresql storage", "Recursive read due to Cluster Cache invalidation race condition");
```
Comment (adam-yorwerth) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:87]: This should be an error so we notice it

Comment (LorenzoCianciaruso) [reply pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:87]: it's not really an error, right? we're handling the race condition when it happens. I think it's good to have visibility of it, but error would mean that there is something wrong, which is not the case

**Review by adam-yorwerth — APPROVED** [review]

**Pull request merged**